### PR TITLE
use tcp request on truncated dns reply

### PR DESCRIPTION
**Pull request by artemg**
When receiving DNS_ERR_TRUNCATED, make DNS request over TCP.

### CODE REVIEW

**Comment by 98tango** [comment]
We would love to have this TCP support so that we can work with large DNS response. Since there haven't been any updates here in a while, just wondering is this still being worked on?

**Comment by azat** [comment]
Hi @98tango, AFAIK there is no activity on this, but this can be because of https://github.com/libevent/libevent/issues/70#issuecomment-109563233 (@ellzey  ?)

File: evdns.c [hunk 1]
```diff
@@ -2173,6 +2183,84 @@ evdns_request_timeout_callback(evutil_socket_t fd, short events, void *arg) {
...
+    in_len = ntohs(in_len);
+    if( size < in_len + sizeof(in_len)){ // not enough data received
+        return;
+    }
+    if( evbuffer_drain(buf, sizeof(in_len)) )
+        goto err;
+    pp = evbuffer_pullup(buf, in_len);
+    if( !pp )
+        goto err;
+    reply_parse(req->base, pp, in_len);
```
Comment (azat) [inline evdns.c]: req->base should be locked

File: evdns.c [hunk 2]
```diff
@@ -2173,6 +2183,84 @@ evdns_request_timeout_callback(evutil_socket_t fd, short events, void *arg) {
...
+    if( bev )
+        bufferevent_free(bev);
+    return;
+}
+static void evdns_request_event_cb_tcp(struct bufferevent *bev, short what, void *ctx){
+    struct request *req = (struct request *)ctx;
+    if (what & (BEV_EVENT_ERROR | BEV_EVENT_READING | BEV_EVENT_WRITING | BEV_EVENT_TIMEOUT )) {
+        bufferevent_free(bev);
+        if( req ) {
+    	    reply_schedule_callback(req, 0, DNS_ERR_TRUNCATED, NULL);
```
Comment (azat) [inline evdns.c]: req->base should be locked

File: evdns.c [hunk 3]
```diff
@@ -2173,6 +2183,84 @@ evdns_request_timeout_callback(evutil_socket_t fd, short events, void *arg) {
...
+/*   0 ok */
+/*   1 temporary failure */
+/*   2 other failure */
+static int
+evdns_request_transmit_to_tcp(struct request *req, struct nameserver *server) {
+    struct bufferevent *be;
+    uint16_t len;
+    ASSERT_LOCKED(req->base);
+    ASSERT_VALID_REQUEST(req);
+    be = bufferevent_socket_new(req->base->event_base, -1,  BEV_OPT_CLOSE_ON_FREE);
```
Comment (azat) [inline evdns.c]: it is too huge overhead creating tcp for each request, this should be done once (with some number of retries)

File: http.c [hunk 4]
```diff
@@ -1908,6 +1922,7 @@ evhttp_parse_headers_(struct evhttp_request *req, struct evbuffer* buffer)
 			goto error;
 
 		svalue += strspn(svalue, " ");
+		rtrim(svalue);
```
Comment (azat) [inline http.c]: already fixed ac4251976

**Pull request closed without merging**